File: day4.js
//implementation using lists:
function findMissingColors(grid) {
  //list of valid colors:
  let colors = ["🟥", "🟧", "🟨", "🟩", "🟦", "🟪", "🟫"];
  //list of colors:
  let found_colors = [];

  //go through each item, if not in found_colors, add color to list:
  for(let i=0;i<grid.length;i++){
    for(let j=0;j<grid[i].length;j++){
      let currentColor = grid[i][j];
      //adding only if color not in found_colors list:
      if (!found_colors.includes(currentColor)){
        found_colors.push(currentColor);
      }
    }
  }

  //go through and see what colors are not added in list:
  let not_found = [];
  for(let i=0;i<colors.length;i++){
    //indexOf returns -1 only if not found--> add to not_found list:
    if (found_colors.indexOf(colors[i]) ==-1 ){
      not_found.push(colors[i]);
    }
  }
  //return not found
  return not_found  
}

// let input1= [["🟥", "🟧", "🟨", "🟩", "🟦", "🟪", "🟥"],
//   ["🟨", "🟩", "🟦", "🟪", "🟥", "🟧", "🟨"],
//   ["🟦", "🟥", "🟧", "🟨", "🟩", "🟪", "🟦"],
//   ["🟩", "🟦", "🟪", "🟥", "🟧", "🟨", "🟩"],
//   ["🟧", "🟨", "🟩", "🟦", "🟪", "🟥", "🟧"],
//   ["🟪", "🟧", "🟨", "🟩", "🟦", "🟥", "🟪"],
//   ["🟥", "🟦", "🟩", "🟪", "🟨", "🟧", "🟦"]];

// console.log(findMissingColors(input1));

// let input2 = [["🟥", "🟧", "🟨", "🟩", "🟦", "🟥", "🟧"],     
// ["🟨", "🟩", "🟦", "🟥", "🟨", "🟩", "🟦"],     
// ["🟥", "🟧", "🟨", "🟩", "🟦", "🟥", "🟨"],     
// ["🟩", "🟦", "🟥", "🟧", "🟨", "🟩", "🟦"],     
// ["🟨", "🟥", "🟧", "🟨", "🟩", "🟦", "🟥"],     
// ["🟦", "🟩", "🟨", "🟥", "🟧", "🟩", "🟦"],    
// ["🟥", "🟧", "🟨", "🟩", "🟦", "🟨", "🟥"]];

// console.log(findMissingColors(input2));

File: day4.py

def find_missing_colors(grid):
  # establish colors
  colors = ["🟥", "🟧", "🟨", "🟩", "🟦", "🟪", "🟫"]
  # create a dictionary --> map colors to the count
  found_colors = {}
  # for each item in the row, see if exists in dict (if so? add 1: create entry and set to one)
  for row in grid:
    for entry in row:
      #entry already exists:
      if entry in found_colors:
        found_colors[entry] +=1
      #entry doesn't exist:
      else:
        found_colors[entry]= 1
  
  #check to see what colors aren't in the dictionary!
  not_found = []
  for color in colors:
    # color isn't in dictionary--> add to not_found list
    if color not in found_colors.keys():
      not_found.append(color)
  #return all colors not found
  return not_found

# input1=   [["🟥", "🟧", "🟨", "🟩", "🟦", "🟪", "🟥"],
#   ["🟨", "🟩", "🟦", "🟪", "🟥", "🟧", "🟨"],
#   ["🟦", "🟥", "🟧", "🟨", "🟩", "🟪", "🟦"],
#   ["🟩", "🟦", "🟪", "🟥", "🟧", "🟨", "🟩"],
#   ["🟧", "🟨", "🟩", "🟦", "🟪", "🟥", "🟧"],
#   ["🟪", "🟧", "🟨", "🟩", "🟦", "🟥", "🟪"],
#   ["🟥", "🟦", "🟩", "🟪", "🟨", "🟧", "🟦"]]

# input2 = [["🟥", "🟧", "🟨", "🟩", "🟦", "🟥", "🟧"],     
# ["🟨", "🟩", "🟦", "🟥", "🟨", "🟩", "🟦"],     
# ["🟥", "🟧", "🟨", "🟩", "🟦", "🟥", "🟨"],     
# ["🟩", "🟦", "🟥", "🟧", "🟨", "🟩", "🟦"],     
# ["🟨", "🟥", "🟧", "🟨", "🟩", "🟦", "🟥"],     
# ["🟦", "🟩", "🟨", "🟥", "🟧", "🟩", "🟦"],    
# ["🟥", "🟧", "🟨", "🟩", "🟦", "🟨", "🟥"]]

# print(find_missing_colors(input1))
# print(find_missing_colors(input2))